pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+	for w := 0; w < l.config.Workers; w++ {
+		go worker(l, jobs)
+	}
+
+	subnetsDone := 0
+	started := time.Now()
+	discoveryInterval := time.Duration(l.config.DiscoveryInterval) * time.Second
+
+	for {
+		for _, subnet := range subnets {
+			if subnet.network.Contains(subnet.currentIP) {
+				ignored := false
+				currentString := subnet.currentIP.String()
+				for _, ip := range subnet.config.IgnoredIPAddresses {
+					if ip == currentString {
+						ignored = true
+						break
+					}
+				}
+				if ignored {
+					incrementIP(subnet.currentIP)
+					continue
+				}
+				jobSubnet := subnet
+				jobSubnet.currentIP = make(net.IP, len(subnet.currentIP))
+				copy(jobSubnet.currentIP, subnet.currentIP)
+				jobs <- jobSubnet
+				incrementIP(subnet.currentIP)
+			} else {
+				// XXX write the cache a bit more often
+				writeCache(subnet.cacheKey, l.devices[subnet.adIdentifier])
+				subnetsDone++
+			}
+		}
+
+		var delay time.Duration = 0
+		if subnetsDone == len(subnets) {
+			for _, subnet := range subnets {
+				copy(subnet.currentIP, subnet.startingIP)
+			}
+			subnetsDone = 0
+			current := time.Now()
+			elapsed := current.Sub(started)
+			delay = discoveryInterval - elapsed
+		}
+
+		select {
+		case <-l.stop:
+			return
+		case <-time.After(delay):
+			if delay > 0 {
+				started = time.Now()
+			}
+		}
+	}
+}
+
+func (l *SNMPListener) createService(deviceIP string, adIdentifier string, entityID string) {
+	_, present := l.services[entityID]
+	if present {
+		return
+	}
+	svc := &SNMPService{
+		adIdentifier: adIdentifier,
+		entityID:     entityID,
+		deviceIP:     deviceIP,
+		creationTime: integration.Before,
+	}
+	l.services[entityID] = svc
+	l.devices[adIdentifier] = append(l.devices[adIdentifier], deviceIP)
+	l.newService <- svc
+}
+
+func (l *SNMPListener) deleteService(entityID string) {
+	// XXX don't delete on first failure
+	svc, present := l.services[entityID]
+	if present {
+		l.delService <- svc
+		delete(l.services, entityID)
+	}
+}
+
+func incrementIP(ip net.IP) {
+	for i := len(ip) - 1; i >= 0; i-- {
+		ip[i]++
+		if ip[i] > 0 {
+			break
+		}
+	}
+}
+
+// Stop queues a shutdown of SNMPListener
+func (l *SNMPListener) Stop() {
+	l.stop <- true
+}
+
+// GetEntity returns the unique entity ID linked to that service
+func (s *SNMPService) GetEntity() string {
+	return s.entityID
+}
+
+// GetTaggerEntity returns the unique entity ID linked to that service
+func (s *SNMPService) GetTaggerEntity() string {
+	return s.entityID
+}
+
+// GetADIdentifiers returns a set of AD identifiers
+func (s *SNMPService) GetADIdentifiers() ([]string, error) {
+	return []string{s.adIdentifier}, nil
+}
+
+// GetHosts returns the device IP
+func (s *SNMPService) GetHosts() (map[string]string, error) {
+	ips := map[string]string{
+		"": s.deviceIP,
+	}
+	return ips, nil
+}
+
+// GetPorts returns the device ports - currently not supported
+func (s *SNMPService) GetPorts() ([]ContainerPort, error) {
+	return []ContainerPort{}, ErrNotSupported
```

Comment:
I could use the port from config, but I chose to define it statically in the provider instead. Is there another use for it?

Reply:
these are used to resolve `%%port%%` in the check template

Reply:
I'm not sure this is super useful here, as we know the port in advance (same for the whole subnet).